code/modules/language/language_holder.dm

```diff
@@ -0,0 +1,109 @@
+/datum/language_holder
+	var/list/languages = list(/datum/language/common)
+	var/list/shadow_languages = list()
+	var/only_speaks_language = null
+	var/selected_default_language = null
+	var/datum/language_menu/language_menu
+
+	var/omnitongue = FALSE
+	var/owner
+
+/datum/language_holder/New(owner)
+	src.owner = owner
+
+	languages = typecacheof(languages)
+	shadow_languages = typecacheof(shadow_languages)
+
+/datum/language_holder/Destroy()
+	owner = null
+	QDEL_NULL(language_menu)
+
+/datum/language_holder/proc/copy(newowner)
+	var/datum/language_holder/copy = new(newowner)
+	copy.languages = src.languages.Copy()
+	// shadow languages are not copied.
+	copy.only_speaks_language = src.only_speaks_language
+	copy.selected_default_language = src.selected_default_language
+	// language menu is not copied, that's tied to the holder.
+	copy.omnitongue = src.omnitongue
+	return copy
+
+/datum/language_holder/proc/grant_language(datum/language/dt)
+	languages[dt] = TRUE
+
+/datum/language_holder/proc/grant_all_languages(omnitongue=FALSE)
+	for(var/la in subtypesof(/datum/language))
```

Comment:
There's few places you use subtypesof(/datum/language), please cache it, having an all languages list will be useful.